core/lz4/CMakeLists.txt

```
@@ -8,7 +8,23 @@ find_package(LZ4 REQUIRED)
 find_package(xxHash REQUIRED)
 
 target_sources(Core PRIVATE src/ZipLZ4.cxx)
-target_link_libraries(Core PRIVATE xxHash::xxHash LZ4::LZ4)
+# In order to avoid to link libxxHash.a twice to libCore in the final
+# link command to create libCore, we distinguish 3 cases:
+# 1. builtin_lz4 and system xxhash
+#    -> Link lz4 to libCore and make its headers location known
+# 2. builtin_lz4 and builtin_xxhash
+#    -> like 1., but make the location of xxhash known. It's not needed
+#       to explicitly link xxhash since it is linked to lz4
+# 3. system lz4 and system xxhash
+#     -> make the headers and libraries locations known
+if(builtin_lz4)
+  target_link_libraries(Core PRIVATE LZ4::LZ4)
+  if (builtin_xxhash)
+    target_include_directories(Core PRIVATE ${xxHash_INCLUDE_DIR})
+  endif()
+else()
+  target_link_libraries(Core PRIVATE xxHash::xxHash LZ4::LZ4)
+endif()
```

```suggestion
if (builtin_lz4 AND builtin_xxhash)
  target_include_directories(Core PRIVATE ${xxHash_INCLUDE_DIR})
else()
  target_link_libraries(Core PRIVATE xxHash::xxHash)
endif()
target_link_libraries(Core PRIVATE LZ4::LZ4)
```
Still feels wrong.  It feels weird/odd/unexpected (but actual) that despite the `PRIVATE` in:
```
target_link_libraries(lz4 LINK_PRIVATE xxHash::xxHash)
```
CMake is propagating the Hash library onto the link lines of dependent libraries.

---

yes. I do not understand the suggestion. Can you be explicit?

---

The first part is an explicit simplication of the code in the PR.

The second part is a complaint directed at `CMake`.  In my opinion the implementation of `target_link_libraries(... PRIVATE ...)` is broken ( and lead to the problem ).